File: bindings/python/gist/aggregate.py
r"""Result-side aggregation over GIST matches (ADR-352). `search`/`files`/`count` answer *where* a pattern occurs; aggregation answers *how it is distributed* — the question an agent asks next: which files carry the most `TODO`s, which directories concentrate a `panic`, what distinct error codes match `apperr\\.\\w+`, which ADRs the tree cites most. It is a pure post-processing layer over the `Match` records the engine already returns: it never widens `SearchRequest` (the contract stays match-finding-only — presentation and stats are deliberately *not* request options) and never runs a second matcher. from gist import search, tally tally(search("TODO", paths=["services"]), by="dir").top(5) `by` selects the axis — a named one (`"file"` · `"dir"` · `"ext"` · `"match"`) or any `Callable[[Match], str]` for a custom bucketing. Only `MatchKind.MATCH` lines are counted; `-A/-B/-C` context lines are display neighborhood, not matches, so they never inflate a tally."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from posixpath import dirname, splitext
from typing import TYPE_CHECKING

from .request import Match, MatchKind


if TYPE_CHECKING:
    from collections.abc import Iterable, Iterator


# An axis is a function from a match to the bucket label it belongs in.
type GroupKey = Callable[[Match], str]


def by_path(m: Match) -> str:
    """Bucket by the file the match is in."""
    return m.path


def by_dir(m: Match) -> str:
    """Bucket by the match's parent directory (``""`` for a repo-root file)."""
    return dirname(m.path)


def by_extension(m: Match) -> str:
    """Bucket by file extension, dot included (``""`` when the file has none)."""
    return splitext(m.path)[1]


def by_match_text(m: Match) -> str:
    """Bucket by the literal text that matched — the first submatch span, else the stripped line (a match line always carries ≥1 submatch in `--json`). This is the axis for "what distinct tokens did this pattern hit"."""
    return m.submatches[0].text if m.submatches else m.text.strip()


_NAMED_AXES: dict[str, GroupKey] = {
    "file": by_path,
    "dir": by_dir,
    "ext": by_extension,
    "match": by_match_text,
}


def resolve_axis(by: str | GroupKey) -> GroupKey:
    """A named axis (`"file"`/`"dir"`/`"ext"`/`"match"`) or a custom callable. An unknown name is a loud `ValueError` — a typo can't silently pick the wrong grouping and misreport a distribution."""
    if callable(by):
        return by
    try:
        return _NAMED_AXES[by]
    except KeyError:
        options = ", ".join(sorted(_NAMED_AXES))
        msg = f"unknown group axis {by!r}; use one of {options} or a callable"
        raise ValueError(msg) from None


@dataclass(frozen=True, slots=True)
class Group:
    """One aggregation bucket: every match sharing a `key`."""

    key: str
    matches: tuple[Match, ...]

    @property
    def count(self) -> int:
        """Matching lines in this bucket."""
        return len(self.matches)

    @property
    def files(self) -> int:
        """Distinct files this bucket spans (1 for a file-axis bucket, ≥1 for a dir/ext/match axis)."""
        return len({m.path for m in self.matches})


@dataclass(frozen=True, slots=True)
class Tally:
    """Buckets ranked by descending match count (ties broken by key ascending) — the shape a report or an agent reads top-down."""

    groups: tuple[Group, ...]

    @property
    def total(self) -> int:
        """Total matching lines across every bucket."""
        return sum(g.count for g in self.groups)

    @property
    def files(self) -> int:
        """Distinct files across every bucket."""
        return len({m.path for g in self.groups for m in g.matches})

    def top(self, n: int) -> tuple[Group, ...]:
        """The `n` largest buckets (all of them when `n <= 0`)."""
        return self.groups if n <= 0 else self.groups[:n]

    def get(self, key: str) -> Group | None:
        """The bucket labelled `key`, or `None` if the pattern never hit it."""
        return next((g for g in self.groups if g.key == key), None)

    def __iter__(self) -> Iterator[Group]:
        """Iterate buckets in descending-count order."""
        return iter(self.groups)

    def __len__(self) -> int:
        """Number of non-empty buckets."""
        return len(self.groups)


def tally(matches: Iterable[Match], *, by: str | GroupKey = "file") -> Tally:
    """Bucket `matches` along `by` and rank the buckets by descending count. Pure and binary-free: it consumes `Match` records, so it composes with `gist.search(...)` or any other source of them and is unit-testable without the engine. Context lines (`MatchKind.CONTEXT`) are skipped, so a request with `-A/-B/-C` context still tallies only the true matches."""
    axis = resolve_axis(by)
    buckets: dict[str, list[Match]] = {}
    for m in matches:
        if m.kind is MatchKind.MATCH:
            buckets.setdefault(axis(m), []).append(m)
    ordered = sorted(buckets.items(), key=lambda kv: (-len(kv[1]), kv[0]))
    return Tally(tuple(Group(key, tuple(group)) for key, group in ordered))

File: bindings/rust/src/exact/aggregate.rs
//! Result-side aggregation over GIST matches.
//!
//! `search`/`files`/`count` answer *where* a pattern occurs; aggregation
//! answers *how it is distributed* — the question an agent asks next: which
//! files carry the most `TODO`s, which directories concentrate a `panic`, what
//! distinct error codes match `apperr\.\w+`. It is a pure post-processing layer
//! over the [`Match`] records the engine already returns: it never widens
//! [`crate::SearchRequest`] (the contract stays match-finding-only —
//! presentation and stats are deliberately *not* request options) and never
//! runs a second matcher.
//!
//! ```no_run
//! let hot = gist::tally(gist::search("TODO")?, gist::Axis::Dir);
//! for g in hot.top(5) {
//!     println!("{:4}  {}", g.count(), g.key);
//! }
//! # Ok::<(), gist::Error>(())
//! ```
//!
//! [`Axis`] selects the named bucketing; [`tally_by`] takes any
//! `Fn(&Match) -> String` for a custom axis. Only [`MatchKind::Match`] lines are
//! counted — `-A/-B/-C` context lines are display neighborhood, not matches, so
//! they never inflate a tally.

use std::collections::{BTreeSet, HashMap};

use irregex::contract::{Match, MatchKind};

/// The bucketing axis for a [`tally`] — the match property that decides which
/// bucket a match falls into.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Axis {
    /// Bucket by the file the match is in.
    File,
    /// Bucket by the match's parent directory (`""` for a repo-root file).
    Dir,
    /// Bucket by file extension, dot included (`""` when the file has none).
    Ext,
    /// Bucket by the literal text that matched — the first submatch span, else
    /// the stripped line. The axis for "what distinct tokens did this pattern
    /// hit".
    Match,
}

impl Axis {
    /// The bucket label `m` belongs in along this axis.
    #[must_use]
    pub fn key(self, m: &Match) -> String {
        match self {
            Self::File => m.path.clone(),
            Self::Dir => dir_of(&m.path).to_owned(),
            Self::Ext => ext_of(&m.path).to_owned(),
            Self::Match => m
                .submatches
                .first()
                .map_or_else(|| m.text.trim().to_owned(), |s| s.text.clone()),
        }
    }
}

/// Parent directory of a `/`-separated path (`""` for a root-level file). Paths
/// are engine output, always forward-slashed, so this is deliberately not
/// `std::path` (which is platform-separator aware) — matching the Python face's
/// `posixpath.dirname`.
fn dir_of(path: &str) -> &str {
    path.rfind('/').map_or("", |i| &path[..i])
}

/// Extension of a path incl. the leading dot (`""` when the basename has none).
/// A dot in a parent directory never counts, and a leading-dot dotfile
/// (`.gitignore`) has no extension — mirroring `posixpath.splitext`.
fn ext_of(path: &str) -> &str {
    let base = path.rsplit('/').next().unwrap_or(path);
    match base.rfind('.') {
        Some(i) if i > 0 => &base[i..],
        _ => "",
    }
}

/// One aggregation bucket: every match sharing a `key`.
#[derive(Debug, Clone)]
pub struct Group {
    /// The bucket label (the axis value all these matches share).
    pub key: String,
    /// The matches in this bucket, in engine output order.
    pub matches: Vec<Match>,
}

impl Group {
    /// Matching lines in this bucket.
    #[must_use]
    pub fn count(&self) -> usize {
        self.matches.len()
    }

    /// Distinct files this bucket spans (1 for a file-axis bucket, ≥1 for a
    /// dir/ext/match axis).
    #[must_use]
    pub fn files(&self) -> usize {
        self.matches
            .iter()
            .map(|m| m.path.as_str())
            .collect::<BTreeSet<_>>()
            .len()
    }
}

/// Buckets ranked by descending match count (ties broken by key ascending) —
/// the shape a report or an agent reads top-down.
#[derive(Debug, Clone)]
pub struct Tally {
    /// The buckets, largest first.
    pub groups: Vec<Group>,
}

impl Tally {
    /// Total matching lines across every bucket.
    #[must_use]
    pub fn total(&self) -> usize {
        self.groups.iter().map(Group::count).sum()
    }

    /// Distinct files across every bucket.
    #[must_use]
    pub fn files(&self) -> usize {
        self.groups
            .iter()
            .flat_map(|g| g.matches.iter().map(|m| m.path.as_str()))
            .collect::<BTreeSet<_>>()
            .len()
    }

    /// The `n` largest buckets (all of them when `n == 0`).
    #[must_use]
    pub fn top(&self, n: usize) -> &[Group] {
        if n == 0 || n >= self.groups.len() {
            &self.groups
        } else {
            &self.groups[..n]
        }
    }

    /// The bucket labeled `key`, or `None` if the pattern never hit it.
    #[must_use]
    pub fn get(&self, key: &str) -> Option<&Group> {
        self.groups.iter().find(|g| g.key == key)
    }

    /// Number of buckets.
    #[must_use]
    pub fn len(&self) -> usize {
        self.groups.len()
    }

    /// Whether no buckets were formed (nothing was tallied).
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.groups.is_empty()
    }

    /// Borrowing iterator over the buckets, largest first.
    pub fn iter(&self) -> std::slice::Iter<'_, Group> {
        self.groups.iter()
    }
}

impl<'a> IntoIterator for &'a Tally {
    type Item = &'a Group;
    type IntoIter = std::slice::Iter<'a, Group>;
    fn into_iter(self) -> Self::IntoIter {
        self.groups.iter()
    }
}

/// Bucket `matches` along a named [`Axis`] and rank the buckets by descending
/// count. See [`tally_by`] for a custom axis.
///
/// Pure and binary-free: it consumes [`Match`] records, so it composes with
/// [`crate::search`] or any other source and is unit-testable without the
/// engine. Context lines ([`MatchKind::Context`]) are skipped, so a request with
/// `-A/-B/-C` context still tallies only the true matches.
#[must_use]
pub fn tally<I>(matches: I, by: Axis) -> Tally
where
    I: IntoIterator<Item = Match>,
{
    tally_by(matches, |m| by.key(m))
}

/// [`tally`] with a caller-supplied axis: any `Fn(&Match) -> String` decides the
/// bucket label, for groupings the named [`Axis`] set doesn't cover (by symbol,
/// by first path segment, by a captured error code…).
#[must_use]
pub fn tally_by<I, F>(matches: I, key: F) -> Tally
where
    I: IntoIterator<Item = Match>,
    F: Fn(&Match) -> String,
{
    let mut buckets: HashMap<String, Vec<Match>> = HashMap::new();
    for m in matches {
        if m.kind == MatchKind::Match {
            buckets.entry(key(&m)).or_default().push(m);
        }
    }
    let mut groups: Vec<Group> = buckets
        .into_iter()
        .map(|(key, matches)| Group { key, matches })
        .collect();
    // A total order (count descending, then unique key ascending) makes the
    // result deterministic regardless of `HashMap` iteration order.
    groups.sort_by(|a, b| {
        b.matches
            .len()
            .cmp(&a.matches.len())
            .then_with(|| a.key.cmp(&b.key))
    });
    Tally { groups }
}
